File: apps/api/app/routers/conversations.py
from fastapi import APIRouter, Depends, Query, Response, status
from sqlalchemy.orm import Session

from app.dependencies import get_db
from app.schemas.conversation import ConversationCreate, ConversationRead, ConversationUpdate, MessageRead
from app.services.conversation_service import ConversationService

router = APIRouter(prefix="/conversations", tags=["conversations"])
service = ConversationService()


@router.get("", response_model=list[ConversationRead])
def list_conversations(workspace_id: str = Query(...), db: Session = Depends(get_db)) -> list[ConversationRead]:
    return service.list_conversations(db, workspace_id)


@router.post("", response_model=ConversationRead, status_code=status.HTTP_201_CREATED)
def create_conversation(payload: ConversationCreate, db: Session = Depends(get_db)) -> ConversationRead:
    return service.create_conversation(db, payload)


@router.get("/{conversation_id}", response_model=ConversationRead)
def get_conversation(conversation_id: str, db: Session = Depends(get_db)) -> ConversationRead:
    return service.get_conversation_or_404(db, conversation_id)


@router.patch("/{conversation_id}", response_model=ConversationRead)
def update_conversation(
    conversation_id: str,
    payload: ConversationUpdate,
    db: Session = Depends(get_db),
) -> ConversationRead:
    return service.update_conversation(db, conversation_id, payload)


@router.delete("/{conversation_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_conversation(conversation_id: str, db: Session = Depends(get_db)) -> Response:
    service.delete_conversation(db, conversation_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.get("/{conversation_id}/messages", response_model=list[MessageRead])
def list_messages(conversation_id: str, db: Session = Depends(get_db)) -> list[MessageRead]:
    return service.list_messages(db, conversation_id)

File: src/api/conversations.ts
import { apiClient } from "./client";
import type {
  Conversation,
  ConversationCreatePayload,
  ConversationUpdatePayload,
  Message,
} from "@/types/api";

export async function fetchConversations(
  workspaceId: string,
): Promise<Conversation[]> {
  const { data } = await apiClient.get("/conversations", {
    params: { workspace_id: workspaceId },
  });
  return Array.isArray(data) ? data : (data.conversations ?? []);
}

export async function fetchConversation(id: string): Promise<Conversation> {
  const { data } = await apiClient.get(`/conversations/${id}`);
  return data;
}

export async function createConversation(
  payload: ConversationCreatePayload,
): Promise<Conversation> {
  const { data } = await apiClient.post("/conversations", payload);
  return data;
}

export async function updateConversation(
  id: string,
  payload: ConversationUpdatePayload,
): Promise<Conversation> {
  const { data } = await apiClient.patch(`/conversations/${id}`, payload);
  return data;
}

export async function deleteConversation(id: string): Promise<void> {
  await apiClient.delete(`/conversations/${id}`);
}

export async function fetchMessages(
  conversationId: string,
): Promise<Message[]> {
  const { data } = await apiClient.get(
    `/conversations/${conversationId}/messages`,
  );
  return Array.isArray(data) ? data : (data.messages ?? []);
}
